src/compiler/parser.ts

```diff
@@ -2649,6 +2694,26 @@ namespace ts {
             return token() === SyntaxKind.CloseParenToken || isStartOfParameter() || isStartOfType();
         }
 
+        function parseJSDocPostfixTypeOrHigher(): TypeNode {

```

Comment:
1. No, `foo?[]` would need to parse inside parseArrayTypeOrHigher. That does seem like it should parse as `Array<foo | null>`.
2. I don't think we should merge functions that would map to individual rules in a grammar.
3. Thanks for the refactor.